clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -808,7 +808,8 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
                 }
                 log.debug("Subscribed to topic(s): {}", Utils.join(topics, ", "));
                 this.subscriptions.subscribe(topics, listener);
-                metadata.setTopics(subscriptions.groupSubscription());
+                this.metadata.needMetadataForAllTopics(true);
```

Comment:
When the user calls subscribe(), we should set a flag in SubscriptionState saying that a rebalance is needed. As of yet, the user's assignment hasn't changed. When the user calls poll(), we check the flag and begin the rebalance. Once the rebalance completes, we set the new assignment and begin fetching. I'm not sure I see a way for any fetches from the old assignment to reach the user.